src/walk.rs

```diff
@@ -36,13 +36,99 @@ enum ReceiverMode {
 
 /// The Worker threads can result in a valid entry having PathBuf or an error.
 #[allow(clippy::large_enum_variant)]
+#[derive(Debug)]
 pub enum WorkerResult {
     // Errors should be rare, so it's probably better to allow large_enum_variant than
     // to box the Entry variant
     Entry(DirEntry),
     Error(ignore::Error),
 }
 
+/// A batch of items to send over a channel.
+struct Batch<T> {
+    items: Arc<Mutex<Option<Vec<T>>>>,
+}
+
+impl<T> Batch<T> {
+    fn new() -> Self {
+        Self {
+            items: Arc::new(Mutex::new(Some(vec![]))),
+        }
+    }
+
+    fn lock(&self) -> MutexGuard<'_, Option<Vec<T>>> {
+        self.items.lock().unwrap()
+    }
+}
+
+impl<T> Clone for Batch<T> {
+    fn clone(&self) -> Self {
+        Self {
+            items: Arc::clone(&self.items),
+        }
+    }
+}
+
+impl<T> IntoIterator for Batch<T> {
+    type Item = T;
+    type IntoIter = std::vec::IntoIter<T>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.lock().take().unwrap().into_iter()
+    }
+}
+
+/// A batch of WorkerResults.
+type ResultBatch = Batch<WorkerResult>;
+
+/// Wrapper that sends batches of items at once over a channel.
+struct BatchSender<T> {
+    batch: Batch<T>,
+    tx: Sender<Batch<T>>,
+}
+
+impl<T> BatchSender<T> {
+    fn new(tx: Sender<Batch<T>>) -> Self {
+        Self {
+            batch: Batch::new(),
+            tx,
+        }
+    }
+
+    /// Check if we need to flush a batch.
+    fn needs_flush(batch: Option<&Vec<T>>) -> bool {
+        match batch {
+            // Limit the batch size to provide some backpressure
+            Some(vec) => vec.len() >= 0x400,
+            // Batch was already taken by the receiver, so make a new one
+            None => true,
+        }
+    }
+
+    /// Add an item to a batch.
+    fn send(&mut self, item: T) -> Result<(), SendError<()>> {
+        let mut batch = self.batch.lock();
+
+        if Self::needs_flush(batch.as_ref()) {
+            drop(batch);
+            self.batch = Batch::new();
+            batch = self.batch.lock();
+        }
+
+        let items = batch.as_mut().unwrap();
+        items.push(item);
+
+        if items.len() == 1 {
```

Comment:
I wonder if it would be better to send over batches after reaching a certain size. That would mean we could know how large to set the initial capacity of the Vec, and possibly avoid contention on the mutex. However, it means receiver threads could end up waiting longer to get results, especially if it takes a while to find more results in the sender threads, so what you have might be better.

Reply:
I just measured the average batch size: **209**.  I don't think we have to try too hard to send larger batches :)

The risk of a minimum batch size is it could stall a very long time if most results are being filtered out.  We could do something like always send the batch after *N* entries are encountered, regardless of whether they're added to the batch.  But I doubt it's worth it.

There isn't very much mutex contention with this design anyway.  It's only between the receiver and at most one sender, and the receiver critical section is extremely short (just `lock().take().unwrap().into_iter()`).  I actually just checked with `perf trace` and the receiver only blocked 136 times over the whole Chromium benchmark (2.1M files).  Each sender blocked between 3-12 times.